src/main/java/com/cloudbees/jenkins/plugins/BitbucketHookReceiver.java

```diff
@@ -51,7 +78,16 @@ public void doIndex(StaplerRequest req) throws IOException {
             LOGGER.log(Level.FINE, "Received commit hook notification : {0}", body);
             JSONObject payload = JSONObject.fromObject(body);
 
-            payloadProcessor.processPayload(payload, req);
+            if ("Bitbucket-Webhooks/2.0".equals(req.getHeader("user-agent"))) {
+                BitbucketEvent bitbucketEvent = new BitbucketEvent(req.getHeader("x-event-key"));
+                BitbucketPayloadProcessor bitbucketPayloadProcessor = payloadProcessorFactory.create(bitbucketEvent);
+                bitbucketPayloadProcessor.processPayload(payload);
+            } else {
+                LOGGER.log(Level.INFO, "Processing old POST service payload");
+                BitbucketPayloadProcessor bitbucketPayloadProcessor =
+                        payloadProcessorFactory.createOldProcessor(new BitbucketEvent("repo:push"));
```

Comment:
This implementation does not support on-prem implementations of Bitbucket server where the payloads (see (https://confluence.atlassian.com/bitbucketserver0512/event-payload-955159824.html) are different. Check out pull-request 58 and try to incorporate some of those changes.